pkg/controller/direct/vmwareengine/privatecloud_controller.go

```diff
@@ -0,0 +1,250 @@
+		gcpClient: client,
+		id:        id,
+		desired:   obj,
+	}, nil
+}
+
+func (m *privateCloudModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type privateCloudAdapter struct {
+	gcpClient *gcp.Client
+	id        *krm.PrivateCloudIdentity
+	desired   *krm.VMwareEnginePrivateCloud
+	actual    *pb.PrivateCloud
+}
+
+var _ directbase.Adapter = &privateCloudAdapter{}
+
+func (a *privateCloudAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting vmwareengine private cloud", "name", a.id)
+
+	req := &pb.GetPrivateCloudRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetPrivateCloud(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting vmwareengine private cloud %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *privateCloudAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating vmwareengine private cloud", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := VMwareEnginePrivateCloudSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &pb.CreatePrivateCloudRequest{
+		Parent:         a.id.Parent().String(),
+		PrivateCloudId: a.id.ID(),
+		PrivateCloud:   resource,
+	}
+	op, err := a.gcpClient.CreatePrivateCloud(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating vmwareengine private cloud %s: %w", a.id.String(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("vmwareengine private cloud %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created vmwareengine private cloud in gcp", "name", a.id)
+
+	status := &krm.VMwareEnginePrivateCloudStatus{}
+	status.ObservedState = VMwareEnginePrivateCloudObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *privateCloudAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating vmwareengine private cloud", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := VMwareEnginePrivateCloudSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	paths := []string{}
+	if desired.Spec.Description != nil && !reflect.DeepEqual(resource.Description, a.actual.Description) {
+		paths = append(paths, "description")
+	}
+	// TODO: spec.managementCluster is not readable from GCP
+	// TODO: spec.networkConfig is has generated fields from GCP
+
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
```

Comment:
(This should have been fixed in scifi). If the paths is empty, we still want to update the KCC object `.status` for first time acquire.

Reply:
Same here. I’ll apply a batch fix to all the controllers I’ve implemented.